File: Java/240. Search a 2D Matrix II.java
/**
 * https://leetcode.com/problems/search-a-2d-matrix-ii/
 */

class Solution {
    // O(n+m) time complexity (n=row, m=column)
    // O(1) space complexity
    public boolean searchMatrix(int[][] matrix, int target) {
        int i = matrix.length - 1;
        int j = 0;

        while (i >= 0 && j < matrix[0].length) {
            if (matrix[i][j] > target)
                i--;
            else if (matrix[i][j] < target)
                j++;
            else
                return true;
        }
        return false;
    }
}

File: Kotlin/240. Search a 2D Matrix II.kt
/**
 * https://leetcode.com/problems/search-a-2d-matrix-ii/
 */

class Solution {
    // O(n+m) time complexity
    // O(1) space complexity
    fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
        var i = matrix[0].size - 1
        var j = 0

        while (i >= 0 && j < matrix.size) {
            when {
                matrix[j][i] == target -> return true
                matrix[j][i] < target -> j++
                else -> i--
            }
        }
        return false
    }
}